Reference Code:

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
(train_images,train_labels), (test_images,test_labels)=tf.keras.datasets.mnist.load_data()
train_images=train_images.reshape(60000,28*28)/255.0
test_images=test_images.reshape((10000,28*28))/255.0

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [ ]:
model=models.Sequential([
    layers.Dense(512, activation='relu',input_shape=(28*28,)),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

In [ ]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

history = model.fit(train_images,train_labels,epochs=10,batch_size=128,validation_data=(test_images,test_labels))

Epoch 1/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 8s 13ms/step - accuracy: 0.8704 - loss: 0.4391 - val_accuracy: 0.9671 - val_loss: 0.1101
Epoch 2/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 10s 13ms/step - accuracy: 0.9715 - loss: 0.0922 - val_accuracy: 0.9757 - val_loss: 0.0770
Epoch 3/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 7s 15ms/step - accuracy: 0.9834 - loss: 0.0550 - val_accuracy: 0.9785 - val_loss: 0.0654
Epoch 4/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 6s 12ms/step - accuracy: 0.9902 - loss: 0.0332 - val_accuracy: 0.9759 - val_loss: 0.0805
Epoch 5/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 11s 14ms/step - accuracy: 0.9910 - loss: 0.0282 - val_accuracy: 0.9797 - val_loss: 0.0672
Epoch 6/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 11s 16ms/step - accuracy: 0.9944 - loss: 0.0179 - val_accuracy: 0.9769 - val_loss: 0.0776
Epoch 7/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 6s 13ms/step - accuracy: 0.9952 - loss: 0.0161 - val_accuracy: 0.9807 - val_loss: 0.0681
Epoch 8/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 7s 16ms/step - accuracy: 0.9969 - loss: 0.0101 - val_a

In [ ]:
test_loss, test_acc=model.evaluate(test_images,test_labels)
print(f'Test accuracy: {test_acc}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9719 - loss: 0.1309
Test accuracy: 0.9758999943733215


To increase Accuracy:

1.   Increase the number of neurons
2.   Decrease the number of hidden layers (Specific (not always))



In [ ]:
model = models.Sequential([
    layers.Dense(1024, activation='relu', input_shape=(28*28,)),  # Increased neurons
    layers.Dense(10, activation='softmax')                         # Output layer
])


In [ ]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

history = model.fit(train_images,train_labels,epochs=10,batch_size=128,validation_data=(test_images,test_labels))

Epoch 1/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 12s 23ms/step - accuracy: 0.8850 - loss: 0.4101 - val_accuracy: 0.9642 - val_loss: 0.1161
Epoch 2/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 11s 24ms/step - accuracy: 0.9706 - loss: 0.1014 - val_accuracy: 0.9755 - val_loss: 0.0830
Epoch 3/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 10s 21ms/step - accuracy: 0.9828 - loss: 0.0574 - val_accuracy: 0.9771 - val_loss: 0.0728
Epoch 4/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 10s 20ms/step - accuracy: 0.9891 - loss: 0.0377 - val_accuracy: 0.9803 - val_loss: 0.0606
Epoch 5/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 11s 22ms/step - accuracy: 0.9926 - loss: 0.0257 - val_accuracy: 0.9781 - val_loss: 0.0724
Epoch 6/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 10s 22ms/step - accuracy: 0.9942 - loss: 0.0202 - val_accuracy: 0.9761 - val_loss: 0.0718
Epoch 7/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 11s 23ms/step - accuracy: 0.9967 - loss: 0.0131 - val_accuracy: 0.9765 - val_loss: 0.0725
Epoch 8/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 20ms/step - accuracy: 0.9971 - loss: 0.0112 - v

In [ ]:
test_loss, test_acc=model.evaluate(test_images,test_labels)
print(f'Test accuracy: {test_acc}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9795 - loss: 0.0746
Test accuracy: 0.9828000068664551


In [ ]:
model = models.Sequential([
    layers.Dense(1024, activation='relu', input_shape=(28*28,)),
    layers.Dense(512, activation='relu'),              # Add a second powerful layer
    layers.Dense(10, activation='softmax')
])


In [ ]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

history = model.fit(train_images,train_labels,epochs=10,batch_size=128,validation_data=(test_images,test_labels))

Epoch 1/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 21s 42ms/step - accuracy: 0.8969 - loss: 0.3526 - val_accuracy: 0.9643 - val_loss: 0.1112
Epoch 2/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 18s 37ms/step - accuracy: 0.9767 - loss: 0.0742 - val_accuracy: 0.9742 - val_loss: 0.0793
Epoch 3/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 21s 38ms/step - accuracy: 0.9859 - loss: 0.0434 - val_accuracy: 0.9749 - val_loss: 0.0828
Epoch 4/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 17s 37ms/step - accuracy: 0.9901 - loss: 0.0303 - val_accuracy: 0.9765 - val_loss: 0.0826
Epoch 5/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 20s 36ms/step - accuracy: 0.9915 - loss: 0.0263 - val_accuracy: 0.9795 - val_loss: 0.0726
Epoch 6/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 18s 39ms/step - accuracy: 0.9934 - loss: 0.0202 - val_accuracy: 0.9822 - val_loss: 0.0651
Epoch 7/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 17s 35ms/step - accuracy: 0.9946 - loss: 0.0154 - val_accuracy: 0.9777 - val_loss: 0.0884
Epoch 8/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 22s 39ms/step - accuracy: 0.9948 - loss: 0.0156 - 

In [ ]:
test_loss, test_acc=model.evaluate(test_images,test_labels)
print(f'Test accuracy: {test_acc}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9766 - loss: 0.1070
Test accuracy: 0.9807000160217285


In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models
import numpy as np
import matplotlib.pyplot as plt

# Load and normalize data
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()
train_images = train_images / 255.0
test_images = test_images / 255.0

# Flatten images
train_images = train_images.reshape((-1, 28*28))
test_images = test_images.reshape((-1, 28*28))

# Build improved model
model = models.Sequential([
    layers.Dense(1024, activation='relu', input_shape=(28*28,), kernel_initializer='he_normal'),
    layers.Dropout(0.3),
    layers.Dense(512, activation='relu', kernel_initializer='he_normal'),
    layers.Dropout(0.3),
    layers.Dense(10, activation='softmax')
])

# Compile the model
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# Train the model
history = model.fit(
    train_images, train_labels,
    epochs=20,
    batch_size=128,
    validation_split=0.1,
    verbose=2
)

# Evaluate the model
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f'Test Accuracy: {test_acc:.6f}')


Epoch 1/20
422/422 - 17s - 41ms/step - accuracy: 0.9196 - loss: 0.2592 - val_accuracy: 0.9728 - val_loss: 0.0904
Epoch 2/20
422/422 - 19s - 46ms/step - accuracy: 0.9669 - loss: 0.1106 - val_accuracy: 0.9763 - val_loss: 0.0779
Epoch 3/20
422/422 - 21s - 49ms/step - accuracy: 0.9759 - loss: 0.0760 - val_accuracy: 0.9785 - val_loss: 0.0724
Epoch 4/20
422/422 - 15s - 36ms/step - accuracy: 0.9794 - loss: 0.0630 - val_accuracy: 0.9812 - val_loss: 0.0713
Epoch 5/20
422/422 - 21s - 50ms/step - accuracy: 0.9829 - loss: 0.0528 - val_accuracy: 0.9827 - val_loss: 0.0631
Epoch 6/20
422/422 - 20s - 47ms/step - accuracy: 0.9853 - loss: 0.0452 - val_accuracy: 0.9805 - val_loss: 0.0645
Epoch 7/20
422/422 - 15s - 36ms/step - accuracy: 0.9873 - loss: 0.0399 - val_accuracy: 0.9833 - val_loss: 0.0622
Epoch 8/20
422/422 - 15s - 35ms/step - accuracy: 0.9895 - loss: 0.0327 - val_accuracy: 0.9843 - val_loss: 0.0638
Epoch 9/20
422/422 - 21s - 49ms/step - accuracy: 0.9891 - loss: 0.0329 - val_accuracy: 0.9803 - 

It's difficult to achieve 99% of accuracy in MNIST Dataset using just Dense Neural Network as it lacks in Pattern Analysis. To improve accuracy more than 98.5%, CNN (Convolutional Neural Network) is used which is sort of upgraded version of Dense NN.